swap/types.go

```diff
@@ -49,3 +52,83 @@ type EmitChequeMsg struct {
 type ConfirmChequeMsg struct {
 	Cheque *Cheque
 }
+
+// Uint256 represents an unsigned integer of 256 bits
+type Uint256 struct {
+	Value *big.Int
+}
+
+var minUint256 = big.NewInt(0)
+var maxUint256 = new(big.Int).Sub(new(big.Int).Exp(big.NewInt(2), big.NewInt(256), nil), big.NewInt(1)) // 2^256 - 1
+
+// NewUint256 creates a Uint256 struct with an initial underlying value of 0
+// no Uint256 should have a nil pointer as its value field
+func NewUint256() *Uint256 {
+	u := new(Uint256)
+	u.Value = minUint256
+	return u
+}
+
+// Uint64ToUint256 creates a Uint256 struct based on the given uint64 param
+// any uint64 is valid as a Uint256
+func Uint64ToUint256(base uint64) *Uint256 {
+	u := NewUint256()
+	u.Value = new(big.Int).SetUint64(base)
+	return u
+}
+
+// Set assigns the underlying value of the given Uint256 param to u, and returns the modified receiver struct
+// returns an error when the result falls outside of the unsigned 256-bit integer range
+func (u *Uint256) Set(value *big.Int) (*Uint256, error) {
+	if value.Cmp(maxUint256) == 1 {
+		return nil, fmt.Errorf("cannot set Uint256 to %v as it overflows max value of %v", value, maxUint256)
+	}
+	if value.Cmp(minUint256) == -1 {
+		return nil, fmt.Errorf("cannot set Uint256 to %v as it underflows min value of %v", value, minUint256)
+	}
+	u.Value = value
+	return u, nil
+}
+
+// Copy sets the underlying value of u to a copy of the given Uint256 param, and returns the modified receiver struct
+func (u *Uint256) Copy(v *Uint256) *Uint256 {
+	valueCopy := new(big.Int).Set(v.Value)
+	u.Value = valueCopy
+	return u
+}
+
+// Cmp calls the underlying Cmp method for the big.Int stored in a Uint256 struct as its value field
+func (u *Uint256) Cmp(v *Uint256) int {
+	return u.Value.Cmp(v.Value)
+}
+
+// Equals returns true if the two Uint256 structs have the same underlying values, false otherwise
+func (u *Uint256) Equals(v *Uint256) bool {
+	return u.Cmp(v) == 0
+}
+
+// Add sets u to augend + addend and returns u as the sum
+// returns an error when the result falls outside of the unsigned 256-bit integer range
+func (u *Uint256) Add(augend, addend *Uint256) (*Uint256, error) {
+	sum := new(big.Int).Add(augend.Value, addend.Value)
+	return u.Set(sum)
+}
+
+// Sub sets u to minuend - subtrahend and returns u as the difference
+// returns an error when the result falls outside of the unsigned 256-bit integer range
+func (u *Uint256) Sub(minuend, subtrahend *Uint256) (*Uint256, error) {
+	difference := new(big.Int).Sub(minuend.Value, subtrahend.Value)
```

Comment:
If we create new `big.Int`s on every arithmetic operation anyway is there an advantage to having this "set receiver to result" api style? I believe the reason for this design is to avoid allocations happening during normal arithmetic operations but since a new `big.Int` is allocated here anyway could we not just allocate an entirely new `Uint256` and have a more "readable" API like in many other big number implementations (where big nums are immutable):
```go
x = NewUint256(bigIntValue)
z = x.Add(y) // x remains unmodified
```

Reply:
as you point out, for something like `z := x.Add(y)` to work (if `x` remains unmodified) we'd need to call `NewUint256` inside of `Add`, which in my opinion is less compliant with SRP than the current version.

i do see how `x, err = x.Add(y)` can look better than `_, err = x.Add(x,y)`. 

for the record: the "set receiver to result" style was modelled after the `math/big` `Int` functions, which was the basis for this design (in the spirit of not re-inventing the wheel).

i don't find it particularly hard to read, but in principle i'm not opposed to changing it.